feat(wiz): make wizard nav items more easily discoverable

seanforyou23 opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Closes https://github.com/patternfly/patternfly-react/issues/1588

Updates Wiz to use more descriptive labels for landmark regions and to allow for buttons as well as anchors for the main nav items. Also adds jest/cypress tests.

It would be good to start considering these accessible names and the overall strategy of how to handle label composition as part of our ongoing content strategy (it's microcopy). I created a new issue tag `microcopy review` to help identify where we need more people to weigh in on the approach in question. @abigaeljamie ping

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-4425.surge.sh

jschuler commented on packages/react-core/src/components/Wizard/WizardNavItem.tsx:
```diff
@@ -26,16 +28,22 @@ export const WizardNavItem: React.FunctionComponent<WizardNavItemProps> = ({
   step,
   onNavItemClick = () => undefined,
-  navItemComponent = 'a'
+  navItemComponent = Button
```
In core they use anchor and this could be a breaking change
https://pf4.patternfly.org/documentation/core/components/wizard

seanforyou23 replied:
Yep, I wanted to use the react workspace to facilitate the conversation on both ends since we can better see the components in action with these interactive examples. If we're good with it we'd definitely need to make corresponding changes in core.

seanforyou23 replied:
Seems like it needs to be configurable between link and button, and maybe favor button as the default since we're typically building SPAs these days. In case the user isn't using a single page app they'll want anchors, and in the case of SPA button seems to carry the most appropriate semantics and behavior as selecting them doesn't currently change the page URL address. We could look into hooking in routing with the nav items, but that's starting to look like a good place to break out into a separate effort.

So yea definitely is missing the anchor support now (I can add this back in with a simple prop to toggle between the two). Does that sound ok? @jschuler

mcoker replied:
If we're supporting both `<a>` and `<button>` here, I don't think that's breaking. Shouldn't be breaking unless we remove support for `a.wizardNavLink`. Though I'm not sure we would use the inline button variation here - we could also just add `border: 0;` (I think that's it?) to the `wizardNavLink` styles to support use of the `<button>` html element.

mcoker replied:
@seanforyou23 opened https://github.com/patternfly/patternfly/pull/3216 if you want to have a look.

seanforyou23 replied:
Made some updates, will remove the temporary inline style soon as we get your core PR merged. Let me know if you see any issues with the styling!

mcoker replied:
Oh I forgot to mention - I noticed this weird behavior in chrome where the focus outline applies to the item's pseudo element (the number), too. We can likely fix that with `:focus-within` on the parent list item. WDYT, is that worth updating?

<img width="177" alt="Screen Shot 2020-06-26 at 9 25 19 AM" src="https://user-images.githubusercontent.com/35148959/85867901-32e07a80-b78f-11ea-9ca9-fcf2056c40dc.png">

seanforyou23 replied:
Hmm good question. Currently the number is announced when you focus the item which seems like it may have been done by design. Would changing to focus-within stop the step numbers from being automatically announced? I think the main thing we'd want to try and preserve is the current behavior in Safari in terms of how the items are announced to SRU

mcoker replied:
I'm not aware of any impacts to screen readers when using `:focus-within`, I believe it just changes the presentation of the focus indicator. I made a couple of demos here if you want to see if there is any sort of effect. The links with "focus within" in the text change the focus indicator so that the number doesn't get its own focus ring.

https://codepen.io/mcoker/pen/pogrRwB - using `:focus-within`
https://codepen.io/mcoker/pen/yLeogaw - just changes the link/button's `outline-style`

seanforyou23 replied:
Yea both of those methods seem to announce the step number just fine. `:focus-within` sounds good to me, good catch!

mcoker commented on packages/react-core/src/components/Wizard/WizardNavItem.tsx:
```diff
@@ -26,16 +28,22 @@ export const WizardNavItem: React.FunctionComponent<WizardNavItemProps> = ({
         onClick={() => onNavItemClick(step)}
-        className={css(styles.wizardNavLink, isCurrent && 'pf-m-current', isDisabled && 'pf-m-disabled')}
+        className={css(
+          btnStyles.modifiers.inline,
```
In general, I think we try to avoid this sort of thing where `<NavItemComponent>` is also a `<Button>`, but we just use the button component styles. Another way could be for us just to add the `<button>` html element support for use with this BEM element.

mcoker commented on packages/react-core/src/components/Wizard/WizardNavItem.tsx:
```diff
@@ -26,16 +28,22 @@ export const WizardNavItem: React.FunctionComponent<WizardNavItemProps> = ({
+          btnStyles.modifiers.inline,
+          styles.wizardNavLink,
+          isCurrent && 'pf-m-current',
+          isDisabled && 'pf-m-disabled'
```
if it's a `button`, seems like we would want to use the `disabled` attr?

jgiardino commented:
These updates are great to see! Thanks for working on this. 

I just had some comments on the use of aria-label for landmark regions. I pasted below one of the screencaps shared in the description. The thing that jumps out at me here is the use of "Content area for ... main":

1. Here are guidelines on use of `<main>`: https://www.w3.org/TR/wai-aria-practices-1.2/#aria_lh_main. A label is recommended when you have more than one, but I'm not sure that would ever be the case. More than one `<main>` should only exist if nested inside a `document` or `application`, which I think is highly unlikely for most apps, or at least most instances of a wizard component. 
2. Refer to the last bullet in this section: https://www.w3.org/TR/wai-aria-practices-1.2/#general-principles-of-landmark-design. They recommend against using "Navigation" in the aria-label of a navigation region, because it's redundant. I would argue that using "Content for" is also redundant with the semantic meaning of the `main` region which is meant to provide the contents of the page. 

My general suggestion is to be very conservative about adding labels to our examples. The Navigation component is an example where we definitely want to include labels in our examples, because we vetted the labels with content strategy and design, and the strings we provide are the same for all instances of that component. But in a case like this, there is a potential risk that placeholder strings we provide in examples are never updated by the consumer. And when this happens, it has the opposite effect of what you're hoping to achieve, because an incorrect label could be very confusing to the user. In other words, if in this case the options are: 1) having a descriptive label is nice but not necessary, 2) having no label is fine, vs 3) having an incorrect label is potentially damaging to the user experience, then the safest option is to not include a label in our examples. Or alternatively, find a way to provide a descriptive label where you eliminate the possibility of it being incorrect...

To do this, a possible alternative solution is to use `aria-labelledby` instead, which references the wizard header text (and maybe also the current step?). This is something that the consumer doesn't need to handle, because we would handle that for them. So in this case, the main region would be announced as "Basic First step main" (which would hopefully make more sense with real wizard contents).  This avoids the risk I mention above while allowing us to include a label. _But_ I also consider this as a nice-to-have given my first point above about `main` not needing a label, and so maybe it isn't worth the extra work to hook this up. 🤷 

For the navigation region, we definitely need a unique label, and I can see it being really helpful to include the wizard header text as part of that label. If this is something we want to include, then I would definitely suggest aria-labelledby instead of aria-label for the navigation region, for similar reasons that I mention above (i.e. that it's really risky to include placeholder text in examples that might get ignored by consumers after copying/pasting the examples). So in this case, aria-labelledby could reference the wizard heading, and be announced as "Basic Steps navigation".

Just to clarify though, I'm not opposed to supporting the ability to provide an aria-label in these components, I'm primarily warning against doing this in examples.

<img width="700" alt="wizard-accessibility-current-1" src="https://user-images.githubusercontent.com/5942899/85176451-d7baff00-b247-11ea-9f83-6e0bbd919243.png" />

seanforyou23 commented:
Good stuff, thx for the feedback everyone! I'll make some updates and push the changes soon as they are ready. Keep you posted!

seanforyou23 commented:
Alright, have another round of updates ready! Added some examples so both cases are clearer. I re-swizzled the way the labels are applied to use aria-labelledby approach, it's working quite well but I later realized a newer requirement for the in-page wizard not to have a header so for in-page wiz there's nothing to point aria-labelledby to out of the box. For these examples, I added the bare example title as aria-label to the `nav` and `main` to emulate what it's like for wiz-in-modal. WDYT?

We should now be able to override all these new props and I've done some cleanup to ensure we're not adding aria attributes in cases where none are needed.

The reason the links were not showing up to screen reader users was because they didn't have href, so for these cases we now throw an error to help prevent users from making this mistake.

I noticed the button in WizardToggle renders to the DOM for the in-page wizard, it's just hidden in the background.. is it used for anything? If not I may add a small check to remove it from the DOM for housekeeping sake.

jgiardino commented:
Thanks for making updates!

I'm finally realizing that `<main>` should only be used in the context of a modal, given the requirement that there should only be one `main` region on a page, unless nested inside an `application` or `document`.  Is it possible to enable that?

I don't remember the discussion or decision behind using "Steps" as the nav region label. It might be worth circling back to when that was introduced into core, as there might be some discussion there, or if not, follow up with content strategy on what's recommended here. For all the other nav regions, we had standardized on strings to use "Global", "Local", etc... so that users could expect a level of consistency in labeling across all the applications. I would think the same applies to the nav region in the wizard.

mcoker commented:
@seanforyou23 https://github.com/patternfly/patternfly/pull/3216 was merged. We still have a weird focus outline issue, but there is a follow up to fix that later (next sprint hopefully) https://github.com/patternfly/patternfly/issues/3269

tlabaj commented on packages/react-core/src/components/Wizard/Wizard.tsx:
```diff
@@ -66,6 +68,12 @@ export interface WizardProps extends React.HTMLProps<HTMLDivElement> {
   startAtStep?: number;
   /** Aria-label for the Nav */
   navAriaLabel?: string;
+  /** Sets aria-labelledby on Nav element */
```
I am wondering if these need to be sentence case...

jschuler commented:
<img width="257" alt="Screen Shot 2020-07-13 at 11 00 04 AM" src="https://user-images.githubusercontent.com/869106/87319741-0b601080-c4f8-11ea-8580-f7cdc69bc06a.png">
the substep style looks different than before (centered vs left-aligned) @mcoker

jschuler commented on packages/react-core/src/components/Wizard/WizardBody.tsx:
```diff
@@ -7,14 +7,26 @@ export interface WizardBodyProps {
+  'aria-labelledby': ariaLabelledBy,
+  mainComponent = 'div'
+}: WizardBodyProps) => {
+  const MainComponent = mainComponent as React.ElementType;
```
should the type be ElementType to begin with so you don't have to cast it?

seanforyou23 replied:
Made the change, but isn't won't it be a breaking change? I was asserting to `ElementType` to improve from `any` but to also try and keep from breaking anyone. Lemme know what you think, happy to do whatever we think is best.

seanforyou23 replied:
Oops, nvm, forgot this is a new prop :)

jschuler commented on packages/react-core/src/components/Wizard/WizardBody.tsx:
```diff
@@ -7,14 +7,26 @@ export interface WizardBodyProps {
-  </main>
-);
+  hasNoBodyPadding = false,
+  'aria-label': ariaLabel = null,
```
don't need to init to null

jschuler commented on packages/react-core/src/components/Wizard/WizardNavItem.tsx:
```diff
@@ -26,18 +28,35 @@ export const WizardNavItem: React.FunctionComponent<WizardNavItemProps> = ({
+  ...rest
 }: WizardNavItemProps) => {
-  const NavItemComponent = navItemComponent as any;
+  const NavItemComponent = navItemComponent as React.ElementType;
```
maybe change the type in the props above

jschuler commented on packages/react-core/src/components/Wizard/WizardNavItem.tsx:
```diff
@@ -17,6 +17,8 @@ export interface WizardNavItemProps {
   onNavItemClick?: (step: number) => any;
   /** Component used to render WizardNavItem */
   navItemComponent?: React.ReactNode;
```
do we support only button and a or anything goes? if the former we could limit it to those 2

jschuler commented on packages/react-core/src/components/Wizard/WizardNavItem.tsx:
```diff
@@ -26,18 +28,35 @@ export const WizardNavItem: React.FunctionComponent<WizardNavItemProps> = ({
-  const NavItemComponent = navItemComponent as any;
+  const NavItemComponent = navItemComponent as React.ElementType;
+
+  if (navItemComponent === 'a' && !href) {
```
might want to add also this check `process.env.NODE_ENV !== 'production'`

tlabaj commented:
@seanforyou23 looks like the core fix went in for the alignment.  Can we rebase this and hopefully merge it.

jschuler commented on packages/react-core/src/components/Wizard/WizardNav.tsx:
```diff
@@ -26,7 +29,11 @@ export const WizardNav: React.FunctionComponent<WizardNavProps> = ({
   return (
-    <nav className={css(styles.wizardNav, isOpen && 'pf-m-expanded')} aria-label={ariaLabel}>
+    <nav
+      className={css(styles.wizardNav, isOpen && 'pf-m-expanded')}
```
is the class available in the styles object?

seanforyou23 replied:
I wasn't previously, but is now! Updated

jschuler commented:
Also, not sure when it broke but looks like the Wizard finished step is not up-to-date anymore?
<img width="700" alt="Screen Shot 2020-07-28 at 4 52 18 PM" src="https://user-images.githubusercontent.com/869106/88720443-dc01f400-d0f2-11ea-9bb2-48d7849d5176.png">
Not sure if you want to fix it in this PR or separately

seanforyou23 commented:
@jschuler good catches, fixed up the emtpy state bits in the finished step example, I guess they've been out of date for a while now. Regarding the expanded class, I updated it to use dynamic style property from react-styles, but then noticed it's never applied because isOpen is always false. Does this need to be cleaned up?

The last thing was that the docs build keep failing with different errors on CI vs locally. Local error attached below.. When I fix the path to something that works like `import { CogsIcon } from '@patternfly/react-icons';` and then run `yarn lint:ts --fix` it reverts back to the further qualified path under dist. Not sure what to do here, as one way fails the lint target and the other fails the docs build. Any advice would be appreciated!

<img width="1265" alt="Screen Shot 2020-07-29 at 10 47 09 AM" src="https://user-images.githubusercontent.com/5942899/88815784-b96eea00-d189-11ea-8d65-81ae6c07173a.png">

jschuler commented:
@seanforyou23 you should be able to set the path to `import CogsIcon from '@patternfly/react-icons/dist/js/CogsIcon';` did that fail previously?

seanforyou23 commented:
Yea it doesn't seem to like that. The import works fine so I just ignored it for now if that's ok.

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-catalog-view-extension@4.6.0
 - @patternfly/react-core@4.34.0
 - @patternfly/react-docs@5.6.22
 - @patternfly/react-inline-edit-extension@4.5.47
 - demo-app-ts@4.25.0
 - @patternfly/react-integration@4.25.0
 - @patternfly/react-table@4.13.14
 - @patternfly/react-topology@4.4.49
 - @patternfly/react-virtualized-extension@4.5.37

Thanks for your contribution! :tada:
